src/server/pkg/obj/obj.go

```diff
@@ -222,8 +222,57 @@ type Client interface {
 	IsIgnorable(err error) bool
 }
 
+type checkedReadCloser struct {
+	io.ReadCloser
+	size  uint64
+	count uint64
+}
+
+func newCheckedReadCloser(size uint64, rc io.ReadCloser) io.ReadCloser {
+	return &checkedReadCloser{
+		ReadCloser: rc,
+		size:       size,
+	}
+}
+
+func (crc *checkedReadCloser) Read(p []byte) (int, error) {
+	count, err := crc.ReadCloser.Read(p)
+	crc.count += uint64(count)
+	if err != nil {
+		if errors.Is(err, io.EOF) && crc.count != crc.size {
+			return count, errors.Errorf("read stream ended after the wrong length, expected: %d, actual: %d", crc.size, crc.count)
+		}
+		return count, err
+	} else if crc.count > crc.size {
+		return count, errors.Errorf("read stream ended after the wrong length, expected: %d, actual: %d", crc.size, crc.count)
+	}
+	return count, nil
+}
+
+type wrapperClient struct {
+	Client
+}
+
+func newWrapperClient(c Client) Client {
+	if c == nil {
+		return nil
+	}
+	return &wrapperClient{Client: c}
+}
+
+func (wc *wrapperClient) Reader(ctx context.Context, name string, offset uint64, size uint64) (io.ReadCloser, error) {
+	// Modify the returned ReadCloser to use a checkedReadCloser which will error
+	// if the stream gives us too few or too many bytes
+	rc, err := wc.Client.Reader(ctx, name, offset, size)
+	if rc != nil {
```

Comment:
This seems like it should be:
```
if err != nil {
    return nil, err
}
return newCheckedReadCloser(size, rc), nil
```

Reply:
I didn't want to enforce any constraints on the wrapped client - if it returns a reader and an error, I would rather leave that behavior untouched.

Reply:
I think that is a constraint we would want to enforce. Unless there is a case where a reader is valid even when the creation of it errors, which would probably be an anti-pattern.

Reply:
Although it's not enforced elsewhere it's really good to have the constraint that methods that return an error and other values return either a nil error or nil values. It's particularly important in cases like this where the non-error value has a `Close` method which probably won't get called if `err != nil`. So I'd vote that we take this is an opportunity to enforce this.

There are some exceptions to this constraint, for example `io.Write` returns an int64 that represents the number of bytes written, it can be non-zero if there's an error after successfully writing some bytes. That case seems meaningfully different from this one though.

Reply:
I get that, but this is a wrapper class, it shouldn't modify the behavior of the underlying client, aside from the one job of the wrapper class which is to add on a check to the reader.

Reply:
There are plenty of places where one function directly returns the `(result, error)` tuple directly from another function, and I think it would be unreasonable to enforce this everywhere.  Adding these kinds of side-effects will make the code harder to follow in the long run.  I think returning a nil result with an error, and a nil error with a result is a very good constraint to have, but this should be implemented in the client class, the wrapper shouldn't make it its responsibility to hide this.

Reply:
That's a fair point, if you don't want to expand the scope of this wrapper I think it's reasonable to not do this. And while we all agree this is an important constraint, it also seems to be one that, afaik, is respected throughout the codebase. At least I don't think we've observed any bugs from this in a long time.

Reply:
This doesn't seem like a good idea to me. The client class should be expected to handle the cases where an error does not invalidate the reader, not the layers above. The expectation should be that they uphold this constraint, and not doing so is a bug in the object storage client (in the case where certain errors are associated with a valid reader). Enabling these semantics just seems like it would lead to problems. I'm not going to hold up this PR for this, and I know there are several clients that will return a non-nil reader when an error occurs, but it does not make sense why we would not start moving in the direction of upholding this constraint (even if the layer below may not).

Reply:
My point is that the wrapper should be a functional, side-effect-free wrapper of the underlying client.  If we add this behavior, it adds extra mental load to understanding why a client is returning what it returns and anyone reading the code needs to be cognizant of this extra layer.  As it stands, the only behavior this wrapper has is that it will return extra errors when there are none, which is very loud and does not require someone working on the code to think about it.  If instead we silently change the return value of the underlying client, we're setting developers up to be surprised when something they think should be happening isn't.  I feel fairly strongly that a wrapper should change the behavior of the wrapped class in as minimal a way as possible, and this includes preserving bugs.

This is very similar to the situation earlier in this PR where you asked me not to silently truncate the returned result from the object client as, while that may let us continue in the case of a bug, it would prevent us from finding and fixing such a bug.  I think that was the right call, but the same principle should be applied here.

Reply:
I don't agree. The wrapper adds the side effect of catching long / short reads, which does not preserve a long / short read bug in the underlying client. I mean, the sole purpose of this wrapper is to catch a bug, no? Returning nil instead of a reader would put a hard stop on a developer attempting to not follow the constraint we want enforced, and they would have to make the appropriate changes in the object storage client to follow the constraint. It just adds extra, valuable checking in my mind.

Reply:
I could get on board with this change if instead of silently discarding a reader we were to raise an error if both a reader and an error were returned, otherwise we are being inconsistent with the design philosophy.

Reply:
That seems fine to me.